### PR TITLE
Extends out the coding principles and adds a coding collection

### PR DESCRIPTION
**Pull request by DannyGB**

### CODE REVIEW

File: collections/_development/code-reviews.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+area: development
+---
+
+# Code Reviews
+
+Code reviews are an essential part of the development process and must be carried out on all code regardless of where the change originates. The process should also be carried out with the appropriate rigour and attention. Some things to keep in mind during the code review process:
+
+* Give reviews your full attention
+* Limit the amount of time you spend reviewing in any one sitting
+* If necessary check the code out and test it locally
```
Comment (dashton82) [inline collections/_development/code-reviews.md]: Personally I think you should always get the branch and run locally. A further step is to make sure it functionally works as per the requirement before reviewing.

File: collections/_development/code-reviews.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+* Give constructive feedback
+* Consider including non-developers in the review process, they may pick up different issues
+* Do not merge code when there are unresolved comments or requests for change
```
Comment (dashton82) [inline collections/_development/code-reviews.md:16]: I think that it can be mentioned that it is the exception that a developer is merging the code back into master. The QA responsible for testing should be responsible for merging the branch back into master after the piece of work has been PO reviewed.

File: principles/coding-principles.md [hunk 3]
```diff
@@ -15,8 +16,10 @@ They are guidance, not The Law - there will always be edge cases, but you should
 
 Correct means provably correct - with tests. All fixes & new features should include tests to prevent regressions.
 Choose clarity over cleverness.
-Don’t Repeat Yourself - The ‘[Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))’ is a good approach to managing duplication. Less code is usually better, but not at the expense of clarity - Clean Coding techniques can be used to ensure the code remains correct, clear and concise.
-Leave things better than you found them - Boyscout the code
+* Don’t Repeat Yourself - The ‘[Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))’ is a good approach to managing duplication. 
+* Less code is usually better, but not at the expense of clarity - Clean Coding techniques can be used to ensure the code remains correct, clear and concise.
+* Leave things better than you found them - Boyscout the code
+* Follow [SOLID principles](https://en.wikipedia.org/wiki/SOLID). This will make code more understandable and easier to maintain
```
Comment (dashton82) [inline principles/coding-principles.md:23]: I'm a big pusher of no comments in code - as mentioned above the code should be clear and concise and also self describing. Comments that are added are never then updated to reflect changes made in further pieces of work.

File: collections/_development/code-reviews.md (hunk 1 above)
Comment (DannyGB) [reply collections/_development/code-reviews.md]: I agree that in the main you should be getting out the code locally and I'll update this to reflect that. However there will always be edge cases (just updating the README springs to mind) so that also needs reflecting in the Principles

File: collections/_development/code-reviews.md (hunk 2 above)
Comment (DannyGB) [reply collections/_development/code-reviews.md:16]: Good call, will add in info around QA and PO involvement in code reviews and the merge process

File: principles/coding-principles.md (hunk 3 above)
Comment (DannyGB) [reply principles/coding-principles.md:23]: I agree, edge cases exist but in the main the code should be self-describing and refactored if it isn't, as per Clean Coding techniques

File: collections/_development/code-branching.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+title:  "Code Branching"
+area: development
+---
+
+# Code Branching
+
+Maintaining a consistent code branching strategy can help to ensure the development process is as smooth as possible and avoid many pitfalls when a team is  working on the same codebase.
+
+### 1. Create a branch from master for each new piece of work
+When the work in the branch is complete create a pull request which can be reviewed. Local-only branches made from the feature branch can be used to split out the work further, these can then be re-integrated back into the feature branch before being pushed to the origin.
```
Comment (CodeWolf86) [inline collections/_development/code-branching.md]: Do you want to mention anything here about a team working on a branch? ie a Feature branch could be shared by a team who then work locally or have their own branches of it with team reviewed PR's?

File: collections/_development/code-branching.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+### 2. Branches should be as short-lived as possible
+So that rebase and merging operations are as straight forward as possible branches should be as short-lived as they can be.
```
Comment (CodeWolf86) [inline collections/_development/code-branching.md]: Repeating the title at the end? maybe tweak to 
"To minimise effort on merging and rebase operations, short lived branched reduce the potential for conflicts to emerge if multiple workstreams move concurrently"

File: collections/_development/code-reviews.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+Code reviews are an essential part of the development process and must be carried out on all code regardless of where the change originates. The process should also be carried out with the appropriate rigour and attention. Some things to keep in mind during the code review process:
+
+* Give reviews your full attention
+* Limit the amount of time you spend reviewing in any one sitting
+* In the main you should check the code out during the review and test it locally
+* Give **constructive** feedback
+* Consider including non-developers in the review process, they may pick up different issues
+* Do not merge code when there are unresolved comments or requests for change
+* * Code merging in the main is a responsibility of the Quality Assurance team and will performed after review from the Product Owner
```
Comment (CodeWolf86) [inline collections/_development/code-reviews.md]: "and will be"

File: collections/_development/tech-debt.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+---
+layout: default
+title:  "Technical Debt"
+area: development
+---
+
+## Technical Debt
+
+[Technical Debt](https://en.wikipedia.org/wiki/Technical_debt) is a concept in software development that reflects the implied cost of additional rework caused by intentionally choosing an easy (limited) solution now instead of using a better approach that would take longer. When this debt is taken on it will likely need paying back at some future point and to be able to effectively manage the debt it is necessary to log and triage it.
```
Comment (CodeWolf86) [inline collections/_development/tech-debt.md:9]: I'm not sure I entirely agree with the paragraph. You're not wrong that in some instances there is an intentional choice of a limited solution. But i'd argue that a number of times, at the point of solution being created, not all requirements that matter are available. Something comes along in 6 months time which wasn't apparent and the way it was originally built doesn't help solve this new problem. This causes rework or workarounds which also gradually introduces tech debt as it is not always feasible to rework every time something comes in or built such an extensible solution up front it makes the initial effort untenable.

File: principles/coding-principles.md [hunk 8]
```diff
@@ -15,11 +16,13 @@ They are guidance, not The Law - there will always be edge cases, but you should
 
 Correct means provably correct - with tests. All fixes & new features should include tests to prevent regressions.
 Choose clarity over cleverness.
-Don’t Repeat Yourself - The ‘[Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))’ is a good approach to managing duplication. Less code is usually better, but not at the expense of clarity - Clean Coding techniques can be used to ensure the code remains correct, clear and concise.
-Leave things better than you found them - Boyscout the code
+* Don’t Repeat Yourself - The ‘[Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))’ is a good approach to managing duplication. 
+* Less code is usually better, but not at the expense of clarity - Clean Coding techniques can be used to ensure the code remains correct, clear and concise.
+* Leave things better than you found them - Boyscout the code
```
Comment (CodeWolf86) [inline principles/coding-principles.md]: "boyscout the code"

File: principles/coding-principles.md [hunk 9]
```diff
@@ -15,11 +16,13 @@ They are guidance, not The Law - there will always be edge cases, but you should
...
+* Follow [SOLID principles](https://en.wikipedia.org/wiki/SOLID). This will make code more understandable and easier to maintain
 
 ### 2. Optimize for change
-Don’t try to solve every conceivable problem up-front, instead focus on making your code easy to change when needed
+Don’t try to solve every conceivable problem up-front, instead focus on making your code easy to change when needed.
 Don’t prematurely optimize - choose clarity over performance, unless there is a serious performance issue that needs to be addressed.
```
Comment (CodeWolf86) [inline principles/coding-principles.md]: Should this line be moved under the below heading?

File: collections/_development/code-branching.md (hunk 5 above)
Comment (DannyGB) [reply collections/_development/code-branching.md]: going for a combination:

`Short-lived branches minimise the effort involved when performing merging & rebase operations and reduce the potential for conflicts to emerge if multiple workstreams move concurrently`

File: collections/_development/code-branching.md (hunk 4 above)
Comment (DannyGB) [reply collections/_development/code-branching.md]: Yes, I'll add something in

File: collections/_development/tech-debt.md (hunk 7 above)
Comment (DannyGB) [reply collections/_development/tech-debt.md:9]: I agree that new or changing requirements do lead to tech-debt, but the decision not to refactor (re-work or apply work-arounds to) the existing code to make the "best" solution is still an intentional choice to take on more tech-debt.

Code degradation over time is also a thing and should be considered under the Boy scout rule

File: principles/coding-principles.md (hunk 9 above)
Comment (DannyGB) [reply principles/coding-principles.md]: Both this heading and the one below are related, I'll re-think the layout and text

**Review by CodeWolf86 — APPROVED** [review]

**Review by satya-sanikommu — APPROVED** [review]
